picocli-codegen/src/main/java/picocli/codegen/annotation/processing/AnnotationValidator.java

```diff
@@ -196,16 +208,33 @@ private void validateInvalidCombinations(RoundEnvironment roundEnv) {
     }
 
     private <T1 extends Annotation, T2 extends Annotation> void validateInvalidCombination(
-            RoundEnvironment roundEnv, Class<T1> c1, Class<T2> c2) {
+        RoundEnvironment roundEnv, Class<T1> c1, Class<T2> c2) {
         for (Element element : roundEnv.getElementsAnnotatedWith(c1)) {
             if (element.getAnnotation(c2) != null) {
                 AnnotationMirror annotationMirror = getPicocliAnnotationMirror(element);
                 error(element, annotationMirror, "%s cannot have both @%s and @%s annotations",
-                        element, c1.getCanonicalName(), c2.getCanonicalName());
+                    element, c1.getCanonicalName(), c2.getCanonicalName());
             }
         }
     }
 
+    private void maybeValidateFinalFields(Element e, TypeMirror type, String annotation) {
+        ElementKind elementKind = e.getKind();
+        boolean isFinal = e.getModifiers().contains(Modifier.FINAL);
+        // We have an existing validator for fields in an interface. Ignore those elements here.
+        boolean shouldValidateFinal = elementKind.isField() && !isInterfaceField(e);
+        if (!isFinal || !shouldValidateFinal) {
+            return;
+        }
+        boolean isConstantValueField = false;
+        for (VariableElement variableElement : ElementFilter.fieldsIn(Collections.singletonList(e))) {
+             isConstantValueField = isConstantValueField || variableElement.getConstantValue() != null;
+        }
+        if ((type.getKind().isPrimitive() || typeUtils.isAssignable(type, stringType)) && isConstantValueField) {
```

Comment:
While I'd love to also apply this compile-time validation to the boxed primitive types, I don't know if there's actually a way to determine if `final Integer foo` has a value specified in the class definition. `getConstantValue()` only returns a non-null value if the defined value is non-null **and** the type is a true primitive type or string.